pkg/operator/ceph/testlib/spec.go

```diff
@@ -0,0 +1,67 @@
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "ceph-op-testlib")
+
+// TestArguments returns a descriptive error if any of the expected arguments do not exist.
+// This supports arguments in which flags appear multiple times with different values but does not
+// support multiple instances of the same flag-value. This test is designed to fail if the list
+// of actual arguments contains extra arguments not specified in the expected args.
+// The expected arguments are given as an array of string arrays. This is to support flags which
+// may have multiple values. Examples:
+// expectedArgs := [][]string{
+//     {"-h"},                              // test for a short flag
+//     {"-vvv"},                            // test for a short flag with value(s) specified
+//     {"-d", "3"},                         // test for a short flag with a value specified
+//     {"--verbose"},                       // test for a --bool flag
+//     {"--name=alex"},                     // test for a --flag=value flag
+//     {"--name", "sam"},                   // test for a --flag with a value after a space
+//     {"--full-name", "sam", "goodhuman"}, // test for a --flag with 2 values separated by spaces
+// }
+func TestArguments(actualArgs []string, expectedArgs [][]string) error {
+	// join all args into a big space-separated arg string so we can use string search on it
+	// this is simpler than a bunch of nested for loops and if statements with continues in them
+	fullArgString := strings.Join(actualArgs, " ")
+	logger.Info("testing that actual args: %s\nmatch expected args:%v", fullArgString, actualArgs)
+	for _, arg := range expectedArgs {
+		// We join each individual argument together the same was as the big string
+		validArgMatcher := strings.Join(arg, " ")
+		if validArgMatcher == "" {
+			return fmt.Errorf("Expected argument %v evaluated to empty string; TestArguments() doesn't know what to do", arg)
+		}
+		matches := strings.Count(fullArgString, validArgMatcher)
+		if matches > 1 {
+			return fmt.Errorf("More than one instance of flag '%s' in: %s; TestArguments() doesn't know what to do",
+				validArgMatcher, fullArgString)
+		} else if matches == 1 {
+			// Remove the instance of the valid match so we can't match to it any more
+			fullArgString = strings.Replace(fullArgString, validArgMatcher, "", 1)
+		} else { // zero matches
+			return fmt.Errorf("Expected argument '%s' missing in: %s\n(It's possible the same arg is in expectedArgs twice.)",
+				validArgMatcher, strings.Join(actualArgs, " "))
+		}
+	}
+	if remainingArgs := strings.Trim(fullArgString, " "); remainingArgs != "" {
+		return fmt.Errorf("The actual arguments have additional args specified: %s", remainingArgs)
+	}
+	return nil
+}
+
+// AddCephCommonDaemonArgsToExpectedArgs appends all of the common Ceph daemon args to expected args
+func AddCephCommonDaemonArgsToExpectedArgs(namespace string, expectedArgs [][]string) [][]string {
```

Comment:
There is a special function for this because of the way `expectedArgs` is formatted. `CephCommonDaemonArgs` simply returns a `[]string` of args. This function returns a `[][]string` where args are grouped with their values. The grouping is to make sure that the args and values aren't separated. An alternative would be to change the way `CephCommonDaemonArgs` returns args to also be paired, but there would have to be some array flattening done after for the call to it in the `spec.go` file, which also might be kind of ugly. This is one of the few things in the code I wrote here that I'm okay with but not happy with; I'd welcome any ideas for changing things to make this cleaner.

Reply:
ok i missed the different return type. Looks reasonable.

Reply:
In getting rid of the `--keyring` and `--config` common ceph daemon options, I saw a good opportunity to fix this.